File: app/src/main/java/com/originb/inkwisenote/common/Logger.java
package com.originb.inkwisenote.common;

import android.util.Log;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.util.ArrayList;
import java.util.List;

public class Logger {

    private String EventCode;

    private final ObjectMapper objectMapper = new ObjectMapper();

    public Logger(String eventCode) {
        this.EventCode = eventCode;
    }

    public Logger forEvent(String eventCode) {
        return new Logger(eventCode);
    }

    public String getDebugInfo() {
        return EventCode;
    }

    public void debug(String message) {
        Log.d(EventCode, message);
    }

    public void debug(String message, Object obj) {
        try {
            List<Object> debugObjects = new ArrayList<>();
            debugObjects.add(message);
            debugObjects.add(obj);

            Log.d(EventCode, objectMapper.writeValueAsString(debugObjects));
        } catch (JsonProcessingException e) {
            debug(message);
            exception("Failed to serialize input object", e);
        }
    }

    public void error(String message) {
        Log.e(EventCode, message);
    }


    public void error(String message, Object obj) {
        try {
            List<Object> debugObjects = new ArrayList<>();
            debugObjects.add(message);
            debugObjects.add(obj);

            Log.e(EventCode, objectMapper.writeValueAsString(debugObjects));
        } catch (JsonProcessingException e) {
            error(message);
            exception("Failed to serialize input object", e);
        }
    }

    public void exception(String message, Throwable exception) {
        Log.e(EventCode, message, exception);
    }
}

File: app/src/main/java/com/originb/inkwisenote2/common/Logger.kt
package com.originb.inkwisenote2.common

import android.util.Log
import com.fasterxml.jackson.core.JsonProcessingException
import com.fasterxml.jackson.databind.ObjectMapper

class Logger(val debugInfo: String?) {
    private val objectMapper = ObjectMapper()

    fun forEvent(eventCode: String?): Logger {
        return Logger(eventCode)
    }

    fun debug(message: String?) {
        Log.d(this.debugInfo, message ?: "null")
    }

    fun debug(message: String?, obj: Any?) {
        try {
            val debugObjects: MutableList<Any?> = mutableListOf(message, obj)
            Log.d(this.debugInfo, objectMapper.writeValueAsString(debugObjects))
        } catch (e: JsonProcessingException) {
            debug(message)
            exception("Failed to serialize input object", e)
        }
    }

    fun error(message: String?) {
        Log.e(this.debugInfo, message ?: "null")
    }


    fun error(message: String?, obj: Any?) {
        try {
            val debugObjects: MutableList<Any?> = mutableListOf(message, obj)
            Log.e(this.debugInfo, objectMapper.writeValueAsString(debugObjects))
        } catch (e: JsonProcessingException) {
            error(message)
            exception("Failed to serialize input object", e)
        }
    }

    fun exception(message: String?, exception: Throwable?) {
        Log.e(this.debugInfo, message, exception)
    }
}
